Water-Physics/Assets/Code/Scripts/Environment/PhysicsHelper.cs

```diff
@@ -0,0 +1,109 @@
+﻿using UnityEngine;
+
+namespace Code.Scripts.Environment
+{
+    public static class PhysicsHelper
+    {
+
+        public static void ApplyForceToReachVelocity(Rigidbody rigidbody, Vector3 velocity, float force = 1, ForceMode mode = ForceMode.Force)
+        {
+            if (force == 0 || velocity.magnitude == 0)
+                return;
+
+            velocity = velocity + velocity.normalized * 0.2f * rigidbody.drag;
+
+            //force = 1 => need 1 s to reach velocity (if mass is 1) => force can be max 1 / Time.fixedDeltaTime
+            force = Mathf.Clamp(force, -rigidbody.mass / Time.fixedDeltaTime, rigidbody.mass / Time.fixedDeltaTime);
+
+            //dot product is a projection from rhs to lhs with a length of result / lhs.magnitude https://www.youtube.com/watch?v=h0NJK4mEIJU
+            if (rigidbody.velocity.magnitude == 0)
+            {
+                rigidbody.AddForce(velocity * force, mode);
+            }
+            else
+            {
+                var velocityProjectedToTarget = (velocity.normalized * Vector3.Dot(velocity, rigidbody.velocity) / velocity.magnitude);
+                rigidbody.AddForce((velocity - velocityProjectedToTarget) * force, mode);
+            }
+        }
+
+        public static void ApplyTorqueToReachRps(Rigidbody rigidbody, Quaternion rotation, float rps, float force = 1)
+        {
+            var radPerSecond = rps * 2 * Mathf.PI + rigidbody.angularDrag * 20;
+
+            float angleInDegrees;
+            Vector3 rotationAxis;
+            rotation.ToAngleAxis(out angleInDegrees, out rotationAxis);
+
+            if (force == 0 || rotationAxis == Vector3.zero)
+                return;
+
+            rigidbody.maxAngularVelocity = Mathf.Max(rigidbody.maxAngularVelocity, radPerSecond);
+
+            force = Mathf.Clamp(force, -rigidbody.mass * 2 * Mathf.PI / Time.fixedDeltaTime, rigidbody.mass * 2 * Mathf.PI / Time.fixedDeltaTime);
+
+            var currentSpeed = Vector3.Project(rigidbody.angularVelocity, rotationAxis).magnitude;
+
+            rigidbody.AddTorque(rotationAxis * (radPerSecond - currentSpeed) * force);
+        }
+
+        public static Vector3 QuaternionToAngularVelocity(Quaternion rotation)
+        {
+            float angleInDegrees;
+            Vector3 rotationAxis;
+            rotation.ToAngleAxis(out angleInDegrees, out rotationAxis);
+
+            return rotationAxis * angleInDegrees * Mathf.Deg2Rad;
+        }
+
+        public static Quaternion AngularVelocityToQuaternion(Vector3 angularVelocity)
+        {
+            var rotationAxis = (angularVelocity * Mathf.Rad2Deg).normalized;
+            float angleInDegrees = (angularVelocity * Mathf.Rad2Deg).magnitude;
+
+            return Quaternion.AngleAxis(angleInDegrees, rotationAxis);
+        }
+
+        public static Vector3 GetNormal(Vector3[] points)
+        {
+            //https://www.ilikebigbits.com/2015_03_04_plane_from_points.html
```

Comment:
Ligt deze link even toe of verwijder hem

Reply:
👌

Reply:
Done